### DISCUSSION TITLE
Add a global helper for de-sugaring rest parameters

### DISCUSSION THREAD

**Discussion by razzp** [babel/babel#12313] [Ideas]
I'm working on a project that includes many functions with rest parameters, ie:

```
function foo(...bar) {
    doSomethingWith(bar);
}
```

Babel transpiles this by injecting per-basis helper code, thus:

```
function foo() {
    for (var _len = arguments.length, bar = new Array(_len), _key = 0; _key < _len; _key++) {
        bar[_key] = arguments[_key];
    }

    doSomethingWith(bar);
}
```

This works, but in my case leads to a lot of *almost* identical code.

Instead, how about creating a new catch-all helper to abstract this logic away?

```
function _restArgs(args, startIndex) {
    for (var len = args.length, out = new Array(len > startIndex ? len - startIndex : 0), key = startIndex; key < len; key++) {
        out[key - startIndex] = args[key];
    }

    return out;
}
```

This could then be applied thus:

```
function foo() {
    var bar = _restArgs(arguments, 0);

    doSomethingWith(bar);
}
```

Or, if the rest parameter is not the first:

```
function foo(normalParam) {
    var bar = _restArgs(arguments, 1);

    doSomethingWith(bar);
}
```

I appreciate this would add needless bloat if there is only a single rest parameter in a project, so perhaps it would have to be used conditionally or as an opt-in feature. Anyway, I'm curious if this has been discussed before and whether there are any other factors I'm overlooking which would make something like this unviable.

**Comment by nicolo-ribaudo** [comment]
The problem is that passing `arguments` to a function is a huge depot in older browsers.
